### DISCUSSION TITLE
What do you use to write frontend tests for your apps and why and how?

### DISCUSSION THREAD

**Discussion by caspahouzer** [appcelerator/titanium_mobile#13366] [General]
The title says it all. I search for the best way to get an app tested via `npm run test` directly in the simulators

**Comment by m1ga** [comment]
Not really using any of them besides the tests in the SDK but I just throw them into the thread:

Mocha:
* https://github.com/tidev/titanium_mobile/tree/master/tests<br/>Runs apps, functions and has some checks where it compares [images]( https://github.com/tidev/titanium_mobile/blob/master/tests/Resources/android/snapshots/listView_header_footer%402x.png) with a [screenshot from the simulator](https://github.com/tidev/titanium_mobile/blob/734c41b56cd63010b269ef6b6e6247da2f9d656c/tests/Resources/ti.ui.listview.test.js#L1388)

Appium:
* https://titaniumsdk.com/guide/Titanium_SDK/Titanium_SDK_How-tos/Debugging_and_Profiling/Build_Test_and_Deploy_a_Titanium_Cross-Platform_App_with_Fastlane.html#set-up-appium-server<br/>you [find the ids](https://discuss.appium.io/uploads/default/original/2X/f/fac323109cbc343bb0ca7127fa2c8a914c841c75.png) and can click them or check whats the text etc. 

Waldo
* https://www.waldo.com/ <br/>You upload an APK and it will run in a simulator and you can click through it and it will record the screens and when the tests run again it will compare the screenshots.

**Reply by caspahouzer** [reply]
Ti mocha is about 5 years old. Is it still usable?

**Reply by m1ga** [reply]
it looks like the tests just use plain mocha without the ti package. ti.mocha was [removed 1 year ago](https://github.com/tidev/titanium_mobile/commit/f300174d3b8d65b09eebdf1bdceca92694dd765f). Never look at all the test files :smile: But the tests are quick easy to write like
```js
const imageView = Ti.UI.createImageView();
imageView.addEventListener('load', function () {
	try {
		should(imageView.image).be.a.String();
		should(imageView.image).eql(Ti.Filesystem.resourcesDirectory + 'Logo.png');
	} catch (err) {
		return finish(err);
	}
	finish();
});
imageView.image = Ti.Filesystem.resourcesDirectory + 'Logo.png';
```
https://github.com/tidev/titanium_mobile/blob/master/tests/Resources/ti.ui.imageview.test.js#L54-L64
that will set an image and check later it its loaded.

The other options more visual but for appium you'll need to write the tests to. Waldo is a bit more limited (from my tests some years ago) but fully visual
